internal/os/os.go

```diff
@@ -1,12 +1,260 @@
+	return &OS{
+		AutomaticallyUpgrade:                true,
+		OsCommit:                            UnknownOsImageId,
+		RequestedOsCommit:                   UnknownOsImageId,
+		HostedImagesURL:                     "",
+		GracefulRebootChannel:               gracefulRebootChannel,
+		WaitTillFinishGracefulRebootChannel: waitTillFinishGracefulRebootChannel,
+	}
+}
+
+func (o *OS) UpdateRpmOstreeStatus()  {
+	cmd := exec.Command("rpm-ostree", "status", "--json")
+	stdout, err := cmd.Output()
+
+	if err != nil {
+		log.Error("Failed to run 'rpm-ostree status'")
+	}
+
+	resUnMarsh := StatusStruct{}
+	if err := json.Unmarshal([]byte(stdout), &resUnMarsh); err != nil{
+		log.Error("Failed to unmarshal json")
+	}
+
+	if len(resUnMarsh.Deployments) == 0 {
+		log.Error("Failed to unmarshal json")
+	}
+
+	// go over the deployments, check which one is the booted one and if it's the required
+	for _, deployment := range resUnMarsh.Deployments{
+		if deployment.Booted {
+			o.OsCommit = deployment.Checksum
+			o.LastSucceededUpgradeTime = time.Unix(int64(deployment.Timestamp),0).String()
+			if o.RequestedOsCommit == UnknownOsImageId{
+				o.LastUpgradeTime = time.Unix(int64(deployment.Timestamp),0).String()
+				o.LastUpgradeStatus = "succeeded"
+			}
+		}
+
+		if deployment.Checksum == o.RequestedOsCommit{
+			o.LastUpgradeTime = time.Unix(int64(deployment.Timestamp),0).String()
+			if !deployment.Booted {
+				o.LastUpgradeStatus = "failed"
+			} else {
+				o.LastUpgradeStatus = "succeeded"
+			}
+		}
+	}
+}
+
+func (o *OS) Init(config models.DeviceConfigurationMessage) error {
+	return o.Update(config)
+}
+
+func (o *OS) Update(configuration models.DeviceConfigurationMessage) error {
+	newOSInfo := configuration.Configuration.Os
+	if newOSInfo == nil {
+		return fmt.Errorf("cannot retrieve configuration info.")
+	}
+
+	if newOSInfo.HostedImagesURL != o.HostedImagesURL {
+		log.Infof("Hosted Images URL has been changed to %s", newOSInfo.HostedImagesURL)
+		err := updateURLInEdgeConfFile(newOSInfo.HostedImagesURL)
+		if err != nil{
+			log.Error("Failed updating file edge.conf")
+			return err
+		}
+		o.HostedImagesURL = newOSInfo.HostedImagesURL
+	}
+
+	if newOSInfo.CommitID != o.RequestedOsCommit {
+		log.Infof("The requested commit ID has been changed to %s", newOSInfo.CommitID)
+		o.RequestedOsCommit = newOSInfo.CommitID
+		cmd := exec.Command("rpm-ostree", "update", "--preview")
+		stdout, err := cmd.Output()
+
+		if err != nil {
+			log.Error("Failed to run 'rpm-ostree update --preview'")
+			return err
+		}
+
+		// TODO check - if there is a multiple commits?
+		if !strings.Contains(string(stdout), newOSInfo.CommitID){
+			log.Error("Failed to run 'rpm-ostree update --preview'")
+			return fmt.Errorf("cannot find the new commit hash. %s", newOSInfo.CommitID)
+		}
+
+		err = updateGreenbootScripts()
+		if err != nil {
+			log.Error("Failed to update greenboot scripts")
+			return err
+		}
+
+		cmd = exec.Command("rpm-ostree", "upgrade")
```

Comment:
This will upgrade to the latest commit available, right? How does it being verified to be the exact commit as requested?

Reply:
In the previous check

Reply:
I can use rpm-ostree deploy + commitID 
but it might cost a lot of traffic if we have a lot of devices so I preferred to check by using the "preview" cmd